clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -133,6 +138,9 @@ static ConstAtom *analyzeAllocExpr(Expr *E, Constraints &CS, QualType &ArgTy) {
   else
     Exprs.insert(E);
 
+  if (!FuncName.compare("calloc"))
+    ret = CS.getNTArr();
+
```

Comment:
I don't think you want to process `calloc`'s arguments in the same way as `malloc` or `realloc` which is what it seems like you are doing. The first argument to `calloc` is a *count* (e.g., `n`), and the second is a *size* (e.g., `sizeof(int)`). For `malloc` or `realloc`, these are combined (e.g., `n * sizeof(int)`). The code that's there now is assuming the two things are combined, and tries to break them apart. But for `calloc` we don't need to do that, since we already know we are allocating more than one thing (probably), and thus can return a (NUL-terminated) array. Or, if we want to be sure, we can make sure that *count* is not the constant `1`. Does that make sense?

Reply:
Yes. I think I'm being overly worrisome here, but my idea was if we had something like `calloc(1, 3*sizeof(int))`, we'd want to process that sizeof call properly. If I'm correct, I think the only thing separating your design and mine is that in mine, the expression `sizeof(...` gets processed in the for loop at the very bottom of the function, but in yours it doesn't (?) because we break out early by just returning an `_Nt_array_ptr`. Either way, I think I can make this change without too much changing.

Reply:
But the thing I'm not understanding is that you are always going to return `NTARR` for `calloc`. Why go through all the work of doing so? If you have `calloc(1,sizeof(int)*3)` or if you have `calloc(1,sizeof(int))` your code is going to return `NTARR` isn't it? Ah, perhaps if it doesn't see a `sizeof` it won't. But in that case it shouldn't matter: You can still default to this being used as an array; you can look to the of `E` itself to figure out what the `ArgTy` should be.

Reply:
Ok, I can get behind this. So for calloc, we won't ever return nullptr. We'll set the ArgTy and then return an `_Nt_array_ptr`, and for realloc, we can process the second argument the same as malloc, but we need to have the first argument flow into the LHS of the assignment.

Reply:
Yes, I'd give this a whirl. Quick thought: Hunt around the internet for example uses of `calloc` maybe via Github search(?) and see how often it's used to allocate a single thing, and also see what the patterns look like. Look at `vsftpd` and the other benchmark programs for the same. Then imagine how your change will go.